x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/shared/FrozenCacheService.java

```diff
@@ -759,36 +760,42 @@ private void throwAlreadyEvicted() {
-                            try {
-                                ensureOpen();
-                                final long start = gap.start();
-                                assert regionOwners[sharedBytesPos].get() == CacheFileRegion.this;
-                                writer.fillCacheRange(
-                                    fileChannel,
-                                    physicalStartOffset() + gap.start(),
-                                    gap.start(),
-                                    gap.end() - gap.start(),
-                                    progress -> gap.onProgress(start + progress)
-                                );
-                                writeCount.increment();
-                            } finally {
-                                decRef();
+                    try {
+                        executor.execute(new AbstractRunnable() {
+
+                            @Override
+                            protected void doRun() throws Exception {
+                                if (CacheFileRegion.this.tryIncRef() == false) {
+                                    throw new AlreadyClosedException("Cache file channel has been released and closed");
+                                }
+                                try {
+                                    ensureOpen();
+                                    final long start = gap.start();
+                                    assert regionOwners[sharedBytesPos].get() == CacheFileRegion.this;
+                                    writer.fillCacheRange(
+                                        fileChannel,
+                                        physicalStartOffset() + gap.start(),
+                                        gap.start(),
+                                        gap.end() - gap.start(),
+                                        progress -> gap.onProgress(start + progress)
+                                    );
+                                    writeCount.increment();
+                                } finally {
+                                    decRef();
+                                }
+                                gap.onCompletion();
                             }
-                            gap.onCompletion();
-                        }
 
-                        @Override
-                        public void onFailure(Exception e) {
-                            gap.onFailure(e);
-                        }
-                    });
+                            @Override
+                            public void onFailure(Exception e) {
+                                gap.onFailure(e);
+                            }
+                        });
+                    } catch (Exception e) {
+                        logger.error(() -> new ParameterizedMessage("unexpected exception when submitting task to fill gap [{}]", gap), e);
+                        assert false : e;
+                        gap.onFailure(e);
```

Comment:
If the thread pool takes care of calling `AbstractRunnable.onRejection()` when it fails to submit a task for execution (see #85217) then this code can be removed.